### PR TITLE
[Routing] Fixed unexpected 404 NoConfigurationException

### PR DESCRIPTION
**Pull request by yceruto**
| Q             | A
| ------------- | ---
| Branch?       | 4.2
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #31199 
| License       | MIT

This is the patch for 4.2+
We need a different patch for 3.4 that is more complex, I think.

### CODE REVIEW

File: src/Symfony/Component/Routing/Matcher/UrlMatcher.php [hunk 1]
```diff
@@ -186,14 +186,10 @@ protected function matchCollection($pathinfo, RouteCollection $routes)
...
             $hasRequiredScheme = !$route->getSchemes() || $route->hasScheme($this->context->getScheme());
-            if ($requiredMethods) {
-                if (!\in_array($method, $requiredMethods)) {
-                    if ($hasRequiredScheme) {
-                        $this->allow = array_merge($this->allow, $requiredMethods);
-                    }
+            if ($hasRequiredScheme && $requiredMethods && !\in_array($method, $requiredMethods)) {
+                $this->allow = array_merge($this->allow, $requiredMethods);
 
-                    continue;
```
Comment (yceruto) [inline src/Symfony/Component/Routing/Matcher/UrlMatcher.php]: this `continue` outside the `if ($hasRequiredScheme)` was preventing to collect the `$this->allowSchemes` bellow

**Comment by yceruto** [comment]
(Travis failures are unrelated)

**Review by nicolas-grekas — APPROVED** [review]
LGTM for 4.2 thanks.
For 3.4, what's your advice?

**Comment by yceruto** [comment]
For 3.4 we need to collect `allowSchemes` in a different way (also different places) because the code as of 4.2 has been completely refactored. I still haven't found the right way to do it.

**Comment by yceruto** [comment]
@nicolas-grekas I just rebased to make tests pass, but I lost your changes, I'm sorry :( please, feel free to push them again, thanks!

**Comment by nicolas-grekas** [comment]
> For 3.4 we need to collect allowSchemes in a different way

Or we can forget about it - 4.2 has new features that allow making the difference - 3.4 doesn't. Makes sense?

I added back my changes.

**Comment by nicolas-grekas** [comment]
Thank you @yceruto.

**Comment by yceruto** [comment]
> > For 3.4 we need to collect allowSchemes in a different way

> Or we can forget about it - 4.2 has new features that allow making the difference - 3.4 doesn't. Makes sense?

It's fine to me, thanks for merging.

**Pull request merged**